File: development/lib/development_web/components/headless/radio.ex
defmodule DevelopmentWeb.Components.Headless.Radio do
  @moduledoc """
  Headless **radio** — a single radio control (label + native input + indicator), Base UI parity.

  Wraps a native `<input type="radio">` (native grouping, keyboard and form submission) and exposes
  Base UI's styling surface as **live data-attributes** via the `Radio` hook: the root and the
  `indicator` carry `data-checked` / `data-unchecked` (kept in sync as the selection moves between
  radios of the same `name`), plus `data-disabled` / `data-readonly` / `data-required` from the props.
  `readonly` is emulated (native radios ignore the attribute) by cancelling the click.

  Use several with a shared `name` (a native group) or compose inside `radio_group` / `field`. Style
  the `chelekom-radio*` classes and the `data-*` hooks — ships **no** styling.

  WAI-ARIA APG: https://www.w3.org/WAI/ARIA/apg/patterns/radio/
  """
  use Phoenix.Component

  @doc type: :component
  attr :id, :string, required: true, doc: "Id of the native input (the label derives {id}-root)"
  attr :name, :string, default: nil, doc: "Shared name forms the native radio group"
  attr :value, :string, default: nil, doc: "This radio's submitted value"
  attr :checked, :boolean, default: false, doc: "Whether this radio is selected"
  attr :disabled, :boolean, default: false, doc: "Ignore interaction (data-disabled)"
  attr :readonly, :boolean, default: false, doc: "Block selection but stay focusable (data-readonly)"
  attr :required, :boolean, default: false, doc: "Require a selection for form submit (data-required)"
  attr :class, :any, default: nil, doc: "Extra classes for the root"
  attr :rest, :global

  slot :inner_block, required: true, doc: "The radio label"

  def radio(assigns) do
    ~H"""
    <label
      id={"#{@id}-root"}
      phx-hook="Radio"
      class={["chelekom-radio", @class]}
      data-checked={@checked}
      data-unchecked={!@checked}
      data-disabled={@disabled}
      data-readonly={@readonly}
      data-required={@required}
      {@rest}
    >
      <input
        type="radio"
        id={@id}
        name={@name}
        value={@value}
        checked={@checked}
        disabled={@disabled}
        required={@required}
        aria-readonly={@readonly && "true"}
        data-part="input"
        class="chelekom-radio__input"
      />
      <span
        data-part="indicator"
        class="chelekom-radio__indicator"
        aria-hidden="true"
        data-checked={@checked}
        data-unchecked={!@checked}
      >
      </span>
      <span data-part="label" class="chelekom-radio__label">{render_slot(@inner_block)}</span>
    </label>
    """
  end
end

File: priv/assets/js/radio.js
// Radio — headless single radio (Base UI parity).
//
// Wraps a native `<input type="radio">` and exposes Base UI's styling surface as live data-attributes
// on the root (the `<label>`) and the `[data-part="indicator"]`: `data-checked` / `data-unchecked`
// (mutually exclusive), kept in sync with the input's `checked` state — including when a SIBLING radio
// in the same `name` group steals the selection (the browser unchecks siblings without firing their
// `change`, so each radio also watches the document for same-name changes). `data-disabled` /
// `data-readonly` / `data-required` are rendered server-side. `readonly` is emulated by cancelling the
// click (native radios ignore the `readonly` attribute).
//
// No-JS fallback: the server already renders the correct initial `data-checked`/`data-unchecked`, so
// SSR and controlled (`<.form>`-driven) usage look right before/without the hook.

const Radio = {
  mounted() {
    this.input = this.el.querySelector('[data-part="input"]');
    this.indicator = this.el.querySelector('[data-part="indicator"]');
    if (!this.input) return;

    this.sync();

    this.onClick = (e) => {
      if (this.el.hasAttribute("data-readonly") || this.el.hasAttribute("data-disabled")) e.preventDefault();
    };
    this.onChange = () => this.sync();
    // A sibling becoming checked unchecks this one WITHOUT firing its `change`, so re-sync on any
    // same-name radio change anywhere in the document.
    this.onDocChange = (e) => {
      const t = e.target;
      if (t && t !== this.input && t.type === "radio" && this.input.name && t.name === this.input.name) {
        this.sync();
      }
    };

    this.input.addEventListener("click", this.onClick);
    this.input.addEventListener("change", this.onChange);
    this.input.ownerDocument.addEventListener("change", this.onDocChange, true);
  },

  destroyed() {
    if (this.onDocChange) this.input.ownerDocument.removeEventListener("change", this.onDocChange, true);
  },

  sync() {
    const on = this.input.checked;
    this.el.toggleAttribute("data-checked", on);
    this.el.toggleAttribute("data-unchecked", !on);
    if (this.indicator) {
      this.indicator.toggleAttribute("data-checked", on);
      this.indicator.toggleAttribute("data-unchecked", !on);
    }
  },
};

export default Radio;
